src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -0,0 +1,234 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+
+namespace System.Text.Json.Serialization
+{
+    public sealed class JsonSerializerOptions
+    {
+        internal const int BufferSizeUnspecified = -1;
+        internal const int BufferSizeDefault = 16 * 1024;
+
+        private ClassMaterializer _classMaterializerStrategy;
+        private int _defaultBufferSize = BufferSizeUnspecified;
+        private bool _hasRuntimeCustomAttributes;
+
+        private static readonly GlobalAttributeInfo s_globalAttributeInfo = new GlobalAttributeInfo();
+
+        private static readonly ConcurrentDictionary<ICustomAttributeProvider, object[]> s_reflectionAttributes = new ConcurrentDictionary<ICustomAttributeProvider, object[]>();
+        private readonly Lazy<ConcurrentDictionary<ICustomAttributeProvider, List<Attribute>>> _runtimeAttributes = new Lazy<ConcurrentDictionary<ICustomAttributeProvider, List<Attribute>>>();
```

Comment:
Is this class meant to be thread-safe when adding runtime attributes? If so shouldn't you use a thread-safe collection instead of `List`?

Reply:
Yes. Thanks. For now I added a todo like this:
```cs
        // todo: determine threading guarantees on the List<Attribute>.
        // Most likely a single thread is adding a new poco type and its attributes, so collisions are not likely.
        // We could change List<> to thread-safe version if necessary.
```